### PR TITLE
Add generic license list

### PR DESCRIPTION
**Pull request by aiuto**
Adds the legacy list of license types from Bazel.

Closees #1

### CODE REVIEW

**Comment by aiuto** [comment]
Questions for review.
- Should we have by_exception_only?
- I don't particularly like restricted_if_statically_linked. I would prefer to say "requires_relinkable_library", or something else that states the condition you must fulfill, rather than the prohibition you are likely to have.

There is no rush on this, we can debate forever in the issue or this PR. I just happened to find the work I had done months ago and wanted to get it out there.

**Comment by danielmachlab** [comment]
> Should we have by_exception_only?

I think we should include it because, although `by_exception_only` may be needed more rarely, it is a generic licesnse that would be useful to many open source projects or organizations that decide to depend on these `license_kind` declarations (even though private organizations will probably define their own). Additionally, it seems that there are cases where it is useful to combine `restricted` and `by_exception_only`: https://opensource.google/docs/thirdparty/licenses/#combinations-of-restricted-and-byexceptiononly-code

> I don't particularly like restricted_if_statically_linked. I would prefer to say "requires_relinkable_library", or something else that states the condition you must fulfill, rather than the prohibition you are likely to have.

You make a good point here about the value of stating the condition that must be fulfilled instead of the prohibition you are likely to have. The answer in my mind could really go either way or even to a third alternative. It seems that `restricted_if_statically_linked` implies `restricted_unless_dynamically_linked`, where `restricted_unless_dynamically_linked` states the condition that must be fulfilled. Probably a good idea to get a third opinion on this question if possible.

File: licenses/generic/BUILD [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+# https://github.com/bazelbuild/bazel/blob/fcfca6157b6de903ab942cef2fc460bf8c8da6ed/src/main/java/com/google/devtools/build/lib/packages/License.java#L59
+#
+# Packages may create license rules that use these license_kinds when the
+# package has provided a LICENSE file, but the text in it  does not match any
+# of the well known licenses in @rules_license//licenses/spdx:*
+
+
+load("@rules_license//rules:license_kind.bzl", "license_kind")
+
+package(default_visibility = ["//visibility:public"])
```
[the remark this answers is not in the document]
Comment (aiuto) [reply licenses/generic/BUILD:28]: Done

**Comment by aiuto** [comment]
> Probably a good idea to get a third opinion on this question if possible.

Yes. I could start a thread with our compliance team. You could do the same at your company. 

Another way to phase the condition might be 'requires_dynamic_link_for_distribution'.    That gets rid of the word "restricted", which is a very loose one.  Restricted to what? From using it in your company? From use in the EU? From export?

That said, 'distribution' is still a little too loose. Does it mean distribution to anyone else but the person who compiles it?  Does it mean people outside your company?  Does it mean paid customers? Does it mean someone can access it over the internet, even it if is wholly running on your own server.  I think it is important to steer clear of things that could be confused with the AGPL taint.

But this brings us back to maybe the very ambiguous 'restricted' is more useful in practice.  We use restricted to mean that you can not ship without getting approval from your compliance team.  Each org can set their own policy on various restricted_* conditions. So some can be auto-allowed.

Perhaps the right thing to do is not add this in this PR. Only do the 'none', 'notice', 'unencumbered', which have clear meanings.  Then we can have a long and leisurely discussion of the more difficult cases.  Any conditions we add we have to live with forever, so I don't mind taking as much time as needed to create each.

**Comment by aiuto** [comment]
Proposal: Do the very clear ones now, then do proposal for others in #16

**Comment by danielmachlab** [comment]
You make some very good points. I think your proposal is the right move. I'll start a thread on my end to see if there are any other suggestions or ideas to consider.

**Comment by danielmachlab** [comment]
@aiuto what are a couple example licenses that would fall under this category, `restricted_if_statically_linked`? Seems to me that LGPL would be one, is this the case?

**Comment by aiuto** [comment]
I removed the restricted ones.

Yes. `restricted_if_statically_linked` is the LGPL case.  There may be others, but I have not read enough licenses to know of them.

I just had an interesting idea for enforcement.
- Let's say we have //third_party/libusb:license. 
- It is of kind LGPL, with the condition `restricted_if_statically_linked`
- and the targets  :libusb_static and :libusb_dynamic.
- both have applicable_licenses=//third_party/libusb:license
- libusb_dynamic adds //third_party/libusb:license_dynamic.
- license_dynamic is of `license_kind` 
   @rules_pkg//licenses:generic:dynamic_link, which has `condition` `is_dynamically_linked`
- similarly, libusb_static would have a 'static' path.

A compliance tool would see that 
- //third_party/rules_pkg brings in the condition `restricted_if_statically_linked`
- that other licenses in the package bring in the conditions `is_dynamically_linked` and or `is_statically_linked`. 
- this can all be reasoned about in the base compliance checking because the conditions all come from generic.

Scenarios like this are even more reason to be slow it what we add to generic.

**Comment by danielmachlab** [comment]
Hi Tony, this is an interesting scenario. In this case, the compliance tool would see that the `:libusb_dynamic` target is `restricted_if_statically_linked` but then see that it is `dynamically_linked`, thus not restricted. Additionally, the compliance tool would conclude that `:libusb_static` is "restricted" due to the condition `restricted_if_statically_linked` and the confirmation that it is "statically_linked" by the condition applied `is_statically_linked`. Is my understanding correct?

It would be nice for the compliance tool to determine on its own if a dependency is static or dynamically linked. In the scenario we are discussing, the compliance tool could be fed out-dated information if the dependency is change from static to dynamic linking, or vice-versa, and the conditions applied (`is_statically_linked`/`is_dynamically_linked`) are not kept in sync. Unsure how this implementation would look at the moment.

It might be worth adding the scenario you're describing to the rules_license `examples/` directory. Perhaps something I could do if deemed useful.

**Comment by aiuto** [comment]
The example would be useful. I think getting the LGPL case right is worthy of a public discussion. I am not sure if it should be here alone, or a proposal on bazel-dev.  I'm content to defer that for a few weeks until enough scaffolding is in place so people can download from head and then experiment on their own.

**Comment by danielmachlab** [comment]
Sounds like a good plan Tony

**Pull request merged**